### PR TITLE
feat: add dataflowgraph command group

### PR DESCRIPTION
**Pull request by ketkimnaik**
Add az iot ops dataflowgraph command group

- Introduces 4 new commands (apply, show, list, delete) for managing dataflow graph resources under a dataflow profile, following the same patterns as existing dataflow commands

- apply includes client-side validation before hitting the API: node structure (required nodeType, valid values Source/Graph/Destination, at least one Source and at least one Destination), connection references, and endpoint checks (existence, unsupported types, destination-only endpoints used as source, Kafka missing consumerGroupId)

Testing:
- Ran integration test locally
- Added unit and integration test
- Liveendpoints testing

Screenshot for possible errors for apply commands:

image

### CODE REVIEW

File: azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_int.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    list_names = [g["name"] for g in list_result]
+    assert graph_name in list_names
+
+    # APPLY again (update — add a transformation node)
+    updated_nodes = nodes + [
+        {
+            "name": "graph-node",
+            "nodeType": "Graph",
+            "graphSettings": {
+                "registryEndpointRef": dest_ep,
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_int.py]: Done.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            endpoint_obj = self._get_endpoint(endpoint_ref, instance_name, resource_group_name)
+            endpoint_props = endpoint_obj.get("properties", {})
+            endpoint_type = endpoint_props.get("endpointType", "")
+
+            if endpoint_type in _GRAPH_UNSUPPORTED_ENDPOINT_TYPES:
+                raise InvalidArgumentValueError(
+                    f"Source node '{node.get('name')}' references endpoint '{endpoint_ref}' "
+                    f"of type '{endpoint_type}', which is not supported in data flow graphs. "
+                    f"Only MQTT, Kafka, and OpenTelemetry endpoints are supported."
+                )
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py [hunk 3]
```diff
@@ -0,0 +1,777 @@
...
+                resource_group_name=generate_random_string(),
+                dataflow_endpoint_type="Mqtt",
+                host="aio-broker",
+            ),
+            "destination_endpoint": get_mock_dataflow_endpoint_record(
+                dataflow_endpoint_name="myendpoint2",
+                instance_name=generate_random_string(),
+                resource_group_name=generate_random_string(),
+                dataflow_endpoint_type="Kafka",
+            ),
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py:280]: Done

File: azext_edge/edge/params.py [hunk 4]
```diff
@@ -314,6 +314,28 @@ def load_iotops_arguments(self, _):
...
+    with self.argument_context("iot ops dataflowgraph") as context:
+        context.argument(
+            "instance_name",
+            options_list=["--instance", "-i"],
+            help="IoT Operations instance name.",
+        )
+        context.argument(
+            "dataflow_graph_name",
+            options_list=["--name", "-n"],
+            help="DataflowGraph name.",
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/params.py:326]: It should be DataflowGraph. Replacing everywhere else.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+            if not node_type:
+                raise InvalidArgumentValueError(
+                    f"Node '{node_name}' is missing a 'nodeType'."
+                )
+            if node_type not in VALID_NODE_TYPES:
+                raise InvalidArgumentValueError(
+                    f"Node '{node_name}' has invalid nodeType '{node_type}'. "
+                    f"Valid nodeType values are: {', '.join(sorted(VALID_NODE_TYPES))}."
+                )
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:195]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+from ..common import (
+    DATAFLOW_ENDPOINT_TYPE_SETTINGS,
+    DataflowEndpointType,
+)
+from .instances import Instances
+from .reskit import get_file_config
+
+logger = get_logger(__name__)
+
+console = Console()
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        for node in source_nodes:
+            source_settings = node.get("sourceSettings", {})
+            endpoint_ref = source_settings.get("endpointRef", "")
+            if not endpoint_ref:
+                raise InvalidArgumentValueError(
+                    f"Source node '{node.get('name')}' is missing 'sourceSettings.endpointRef'."
+                )
+            endpoint_obj = self._get_endpoint(endpoint_ref, instance_name, resource_group_name)
+            endpoint_props = endpoint_obj.get("properties", {})
+            endpoint_type = endpoint_props.get("endpointType", "")
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_int.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+    # The test relies on at least one MQTT-type endpoint existing in the instance.
+    endpoints = run(f"az iot ops dataflow endpoint list -g {rg} -i {instance}")
+    mqtt_endpoints = [
+        ep for ep in endpoints
+        if ep.get("properties", {}).get("endpointType") in ("Mqtt", "AIOLocalMqtt", "EventGrid", "CustomMqtt")
+    ]
+    if not mqtt_endpoints:
+        pytest.skip(
+            "No MQTT endpoint found in instance — skipping dataflow graph integration test. "
+            "Create at least one MQTT endpoint before running this test."
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_int.py]: Added comment

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+# Endpoint types supported in data flow graphs (explicit allow-list).
+# MQTT, Kafka, and OpenTelemetry are supported.
+# All other types (DataExplorer, DataLakeStorage, FabricOneLake, LocalStorage, etc.) are not.
+_GRAPH_SUPPORTED_ENDPOINT_TYPES = frozenset([
+    MQTT_ENDPOINT_TYPE,
+    KAFKA_ENDPOINT_TYPE,
+    OPENTELEMETRY_ENDPOINT_TYPE,
+])
+
+
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+                raise InvalidArgumentValueError(
+                    f"nodeConnection at index {i} references unknown 'from' node '{from_name}'. "
+                    f"Declared node names: {', '.join(sorted(declared_names))}."
+                )
+            if to_name and to_name not in declared_names:
+                raise InvalidArgumentValueError(
+                    f"nodeConnection at index {i} references unknown 'to' node '{to_name}'. "
+                    f"Declared node names: {', '.join(sorted(declared_names))}."
+                )
+
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:242]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+        declared_names = set()
+        for i, node in enumerate(nodes):
+            node_name = node.get("name", "")
+            node_type = node.get("nodeType", "")
+
+            if not node_name:
+                raise InvalidArgumentValueError(
+                    f"Node at index {i} is missing a 'name'."
+                )
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:186]: done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+
+        # Validate each source node's endpoint
+        for node in source_nodes:
+            source_settings = node.get("sourceSettings", {})
+            endpoint_ref = source_settings.get("endpointRef", "")
+            if not endpoint_ref:
+                raise InvalidArgumentValueError(
+                    f"Source node '{node.get('name')}' is missing 'sourceSettings.endpointRef'."
+                )
+            endpoint_obj = get_endpoint_cached(endpoint_ref)
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:257]: Done

File: azext_edge/edge/_help.py [hunk 13]
```diff
@@ -746,6 +746,110 @@ def load_iotops_help():
...
+              { "from": { "name": "source-mqtt" }, "to": { "name": "dest-otel" } }
+            ]
+          }
+          ```
+
+          The above example defines a fan-out graph: an MQTT source fans out to a Kafka destination
+          and an OpenTelemetry destination. Data flow graphs support only MQTT, Kafka, and
+          OpenTelemetry endpoints. The file can also be the full ARM resource wrapper
+          (properties is auto-extracted). extendedLocation is always auto-populated from --instance
+          and -g and must not be included in the file.
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/_help.py]: checking with dataflow team.

[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/_help.py]: Got confirmation about type and updated it.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+
+        if not source_nodes:
+            raise InvalidArgumentValueError(
+                "The dataflow graph config must contain at least one node with nodeType 'Source'."
+            )
+        if not destination_nodes:
+            raise InvalidArgumentValueError(
+                "The dataflow graph config must contain at least one node with nodeType 'Destination'."
+            )
+
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Updated PR descrption.

File: azext_edge/edge/params.py [hunk 15]
```diff
@@ -314,6 +314,28 @@ def load_iotops_arguments(self, _):
...
+        context.argument(
+            "profile_name",
+            options_list=["--profile", "-p"],
+            help="Dataflow profile name.",
+        )
+        context.argument(
+            "config_file",
+            options_list=["--config-file"],
+            help="Path to a JSON file containing the dataflowgraph properties.",
+        )
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/params.py:337]: This is followed pattern in codebase. It is fine.

File: azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py [hunk 16]
```diff
@@ -0,0 +1,833 @@
...
+        profile_name=profile_name,
+        instance_name=instance_name,
+        resource_group_name=resource_group_name,
+        config_file="config.json",
+        wait_sec=0.1,
+    )
+
+    assert len(mocked_responses.calls) == 4
+    assert result == file_payload
+    request_payload = json.loads(put_response.calls[0].request.body)
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py:364]: The change not needed.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+        node_connections = graph_config.get("nodeConnections", [])
+        for i, conn in enumerate(node_connections):
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:211]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+                raise InvalidArgumentValueError(
+                    f"Source node '{node.get('name')}' references endpoint '{endpoint_ref}' "
+                    f"of type '{endpoint_type}', which is destination-only and cannot be used as a source."
+                )
+
+            # Kafka source endpoints require a consumerGroupId on the endpoint
+            kafka_types = {DataflowEndpointType.EVENTHUB.value, DataflowEndpointType.CUSTOMKAFKA.value,
+                           KAFKA_ENDPOINT_TYPE}
+            if endpoint_type in kafka_types:
+                consumer_group_id = endpoint_props.get("kafkaSettings", {}).get("consumerGroupId", "")
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

**Comment by ketkimnaik** [comment]
Successful integration test run: https://github.com/ketkimnaik/azure-iot-ops-cli-extension/actions/runs/25194954572/job/73897010810

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 19]
```diff
@@ -0,0 +1,323 @@
...
+from .instances import Instances
+from .reskit import get_file_config
+
+console = Console()
+
+# Endpoint types supported in data flow graphs (MQTT family + Kafka family + OpenTelemetry).
+# DataExplorer, DataLakeStorage, FabricOneLake, LocalStorage are not supported.
+_GRAPH_SUPPORTED_ENDPOINT_TYPES = frozenset([
+    # MQTT family
+    DataflowEndpointType.AIOLOCALMQTT.value,
```
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:27]: nit can we add `_` like `AIO_LOCAL_MQTT` to make it more readable, see line 17

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 20]
```diff
@@ -0,0 +1,323 @@
...
+            )
+            return wait_for_terminal_state(poller, **kwargs)
+
+    def _validate_graph_config(  # noqa: C901
+        self,
+        graph_config: dict,
+        instance_name: str,
+        resource_group_name: str,
+    ):
+        VALID_NODE_TYPES = {"Source", "Graph", "Destination"}
```
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: nit: move to the beginning of the file?

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 21]
```diff
@@ -0,0 +1,323 @@
...
+                    f"{', '.join(sorted(_GRAPH_SUPPORTED_ENDPOINT_TYPES))}."
+                )
+            if endpoint_type in _DESTINATION_ONLY_ENDPOINT_TYPES:
+                raise InvalidArgumentValueError(
+                    f"Source node '{node.get('name')}' references endpoint '{endpoint_ref}' "
+                    f"of type '{endpoint_type}', which is destination-only and cannot be used as a source."
+                )
+
+            # Kafka source endpoints require a consumerGroupId on the endpoint
+            kafka_types = {DataflowEndpointType.EVENTHUB.value, DataflowEndpointType.CUSTOMKAFKA.value,
```
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: you can use a fronzenset for `kafka_types` as well

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 22]
```diff
@@ -0,0 +1,323 @@
...
+                kafka_settings = endpoint_props.get("kafkaSettings") or {}
+                consumer_group_id = kafka_settings.get("consumerGroupId", "")
+                if not consumer_group_id:
+                    raise InvalidArgumentValueError(
+                        f"Source node '{node.get('name')}' references Kafka endpoint '{endpoint_ref}', "
+                        f"but the endpoint does not have 'kafkaSettings.consumerGroupId' set. "
+                        f"A consumer group ID is required for Kafka source endpoints."
+                    )
+
+        # Validate each destination node's endpoint
```
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: so we validate destination node, source node, but no graph node validation? it's another valid node type in line 122

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 23]
```diff
@@ -0,0 +1,323 @@
...
+            raise InvalidArgumentValueError(
+                "The dataflow graph config must contain at least one node with nodeType 'Source'."
+            )
+        if not destination_nodes:
+            raise InvalidArgumentValueError(
+                "The dataflow graph config must contain at least one node with nodeType 'Destination'."
+            )
+
+        node_connections = graph_config.get("nodeConnections", [])
+        if not isinstance(node_connections, list):
```
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: `if not isinstance(node_connections, list) or not node_connections`:

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 24]
```diff
@@ -0,0 +1,323 @@
...
+        resource_group_name: str,
+    ) -> dict:
+        """Fetch a dataflow endpoint by name, raising a clear error if not found."""
+        try:
+            endpoint_obj = self.ops_endpoint.get(
+                instance_name=instance_name,
+                resource_group_name=resource_group_name,
+                dataflow_endpoint_name=endpoint_name,
+            )
+        except ResourceNotFoundError:
```
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:345]: what about other error type, we ignore?

File: azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_int.py [hunk 25]
```diff
@@ -0,0 +1,175 @@
+# coding=utf-8
```
Comment (HangyiWang) [inline azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_int.py:1]: all good, just a few thoughts. we could add some negative/edge-case tests for the validation: empty nodes, duplicate node names, invalid endpoint/nodeType, and Kafka source with other stuff. Happy path coverage looks solid.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 26]
```diff
@@ -0,0 +1,323 @@
...
+            endpoint_ref = source_settings.get("endpointRef", "")
+            if not endpoint_ref:
+                raise InvalidArgumentValueError(
+                    f"Source node '{node.get('name')}' is missing 'sourceSettings.endpointRef'."
+                )
+            endpoint_obj = get_endpoint_cached(endpoint_ref)
+            endpoint_props = endpoint_obj.get("properties", {})
+            endpoint_type = endpoint_props.get("endpointType", "")
+
+            if endpoint_type and endpoint_type not in _GRAPH_SUPPORTED_ENDPOINT_TYPES:
```
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:261]: `if not endpoint_type` ? should we rasie error

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 27]
```diff
@@ -0,0 +1,323 @@
...
+    ) -> dict:
+        resource = {}
+        graph_config = get_file_config(config_file)
+        resource["extendedLocation"] = self.instances.get_ext_loc(
+            name=instance_name,
+            resource_group_name=resource_group_name,
+        )
+        resource["properties"] = graph_config
+
+        self._validate_graph_config(graph_config, instance_name, resource_group_name)
```
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: we should validate graph config before calling `get_ext_loc` in line 98 so no API call needed if config is bad

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 28]
```diff
@@ -0,0 +1,323 @@
...
+            to_val = conn.get("to")
+            if not isinstance(from_val, dict) or not from_val.get("name"):
+                raise InvalidArgumentValueError(
+                    f"nodeConnection at index {i} is missing a valid 'from.name' field."
+                )
+            if not isinstance(to_val, dict) or not to_val.get("name"):
+                raise InvalidArgumentValueError(
+                    f"nodeConnection at index {i} is missing a valid 'to.name' field."
+                )
+            from_name = from_val["name"]
```
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:226]: can `{"from": {"name": "A"}, "to": {"name": "A"}}` pass?

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 19 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:27]: These are enum members, not string constants (like line 17) — the naming convention is different, so keeping the existing style is more appropriate here.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 21 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: kafka_types is a local variable used for a single membership check — mutability isn't a concern here, so a plain set is sufficient.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 24 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:345]: Other errors (auth, network, etc.) intentionally propagate with their original SDK messages — only ResourceNotFoundError is intercepted to give a more actionable message pointing to --config-file.

File: azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_int.py (hunk 25 above)
Comment (ketkimnaik) [reply azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_int.py:1]: These are all client-side validation errors that don't require live resources — they're already covered in test_dataflow_graphs_unit.py (empty nodes: line 372, invalid nodeType: line 398, duplicate names: line 406, Kafka missing consumerGroupId: lines 643–669). Int tests cover only happy-path CRUD against a live instance.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 26 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:261]: An endpoint fetched from Azure should always have endpointType since the API enforces it on create — if it's missing, it's corrupt/unexpected data and silently skipping it is worse than surfacing it. We can leave the guard as-is since the API won't return such a case in practice.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 28 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:226]: Better to add this check.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 27 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Makes sense. Done — _validate_graph_config is now called before get_ext_loc, so invalid configs fail fast with no API calls.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 23 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 22 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Good catch. I didn't add it earlier because we didn't have validation for tranform node in dataflow commands. Tried to follow the similar pattern. It is okay o have extra validation so adding it :)

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 20 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done.

**Comment by HangyiWang** [comment]
Could you run int test in the Github thanks

**Review by HangyiWang — DISMISSED** [review]

File: azext_edge/edge/providers/orchestration/common.py [hunk 29]
```diff
@@ -49,6 +49,7 @@
 
 KAFKA_ENDPOINT_TYPE = "Kafka"
 MQTT_ENDPOINT_TYPE = "Mqtt"
+OPENTELEMETRY_ENDPOINT_TYPE = "OpenTelemetry"
```
Comment (digimaun) [inline azext_edge/edge/providers/orchestration/common.py]: Where is this variable used?

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 30]
```diff
@@ -0,0 +1,343 @@
...
+            poller = self.ops.begin_create_or_update(
+                resource_group_name=resource_group_name,
+                instance_name=instance_name,
+                dataflow_profile_name=dataflow_profile_name,
+                dataflow_graph_name=name,
+                resource=resource,
+            )
+            return wait_for_terminal_state(poller, **kwargs)
+
+    def _validate_graph_config(  # noqa: C901
```
Comment (digimaun) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: We should strive not to introduce `# noqa: C901` often (the apply PR also has this), it should be used sparingly.

Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Refactored _validate_graph_config into focused helper methods (_validate_nodes, _validate_node_connections, _validate_source_nodes, _validate_destination_nodes, _validate_graph_nodes) — # noqa: C901 is no longer needed.

File: azext_edge/edge/providers/orchestration/common.py (hunk 29 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/common.py]: Earlier I defined it to use in Dataflowgraph but afterwards changed to use Dataflowendpoint directly. Let me remove it.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 31]
```diff
@@ -0,0 +1,351 @@
...
+                )
+            if not isinstance(to_val, dict) or not to_val.get("name"):
+                raise InvalidArgumentValueError(
+                    f"nodeConnection at index {i} is missing a valid 'to.name' field."
+                )
+            from_name = from_val["name"]
+            to_name = to_val["name"]
+            if from_name == to_name:
+                raise InvalidArgumentValueError(
+                    f"nodeConnection at index {i} is a self-loop: 'from' and 'to' both reference node '{from_name}'."
```
Comment (peolivei2) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:230]: Do you know if the ARM API checks for loops (other than self-loops)?

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 32]
```diff
@@ -0,0 +1,351 @@
...
+
+    def _validate_graph_nodes(self, graph_nodes: list):
+        for node in graph_nodes:
+            graph_settings = node.get("graphSettings", {})
+            if not isinstance(graph_settings, dict):
+                raise InvalidArgumentValueError(
+                    f"Graph node '{node.get('name')}' has an invalid 'graphSettings' field: "
+                    f"expected an object, got {type(graph_settings).__name__}."
+                )
+            if not graph_settings.get("registryEndpointRef"):
```
Comment (peolivei2) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: We should at least validate if the registry endpoint resource exists.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 33]
```diff
@@ -0,0 +1,351 @@
...
+                raise InvalidArgumentValueError(
+                    f"Graph node '{node.get('name')}' is missing 'graphSettings.registryEndpointRef'."
+                )
+            if not graph_settings.get("artifact"):
```
Comment (peolivei2) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Is there a specific format for the artifact reference (like `name:version`)? If so, we should validate here.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 31 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:230]: It seems they have this validation from ARm side:
image

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 33 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Yes — the ARM spec explicitly states the format is <artifact-name>:<version>, so the colon-separator check was added.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 32 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Added this check.

**Comment by ketkimnaik** [comment]
> Could you run int test in the Github thanks

This is the successful run: https://github.com/ketkimnaik/azure-iot-ops-cli-extension/actions/runs/25399682376/job/74495715993 

All the tests have pass. Delete connector cluster and resources step is failing which is not related to the test.

**Review by HangyiWang — DISMISSED** [review]

File: azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py [hunk 34]
```diff
@@ -0,0 +1,1070 @@
...
+
+    assert expected_error_text in exc.value.args[0]
+
+
+# ---------------------------------------------------------------------------
+# delete
+# ---------------------------------------------------------------------------
+
+
+@pytest.mark.parametrize("confirm_yes", [True])
```
Comment (digimaun) [inline azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py]: nit, either this should be removed or extend it with False/None to exercise the bail out path

**Review by digimaun — DISMISSED** [review]

**Review by HangyiWang — APPROVED** [review]

**Review by digimaun — APPROVED** [review]

File: azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py (hunk 34 above)
Comment (ketkimnaik) [reply azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py]: Done
